opencog/cython/PythonEval.cc

```diff
@@ -382,12 +382,19 @@ static bool try_to_load_modules(const char ** config_paths)
 
 static bool already_initialized = false;
 static bool initialized_outside_opencog = false;
+static bool was_initialized_by_us = false;
```

Comment:
?? I think that `initialized_outside_opencog` is supposed to mean `not was_initialized_by_us` so it seems you have two flags that seem to mean the same thing. This is a potentially confusing situation likely to lead to future bugs ...

Reply:
I agree name is misleading, `was_initialized_by_us` is not strictly opposite to `initialized_outside_opencog` because if `was_initialized_by_us` is set to true once it will never be `false`. `initialized_outside_opencog` can be set and unset few times.

Reply:
better name would be was_finalized_by_us. I'll make merge request

Reply:
here https://github.com/opencog/atomspace/pull/2367

Reply:
So ... `initialized_outside_opencog` can be set and unset few times !? But the point was that python can only be initialized once, now, so I don't see how it can be set more than once.  The original intent of this flag was to know who should finalize: us, if we did the initialization, or someone else, if they did the initialization.

Reply:
If python is initialized outside opencog, then global_python_finalize doesn't really do anything, so it is possible to call global_python_initialize again.

Reply:
@linas, do you mean we could forbid re-initialization for both internally an externally initialized `Python`? We have discussed this with @noskill yesterday. It will simplify the code and it is not clear do we really need maintain different behavior for external initialization case.

Reply:
> it is possible to call global_python_initialize again.

?? Python doesn't care who tried to initialize it a second time.  It will fail, no matter who tries to do it...

> not clear do we really need maintain different behavior for external initialization case.

?? Because, if someone else initialized python, outside of the atomspace, then initializing it a second time will fail. So you have to avoid the second initialization, and that is what that flag kept track of.  The two main use cases are: 

From bash prompt:
```
$ python3
Python 3.7.3 
Type "help", "copyright", "credits" or "license" for more information.
>>>  from opencog.atomspace import AtomSpace
```
The above initialize python "outside" of the atomspace; it has already been set up by the time the atomspace is started.  The other use case is:
```
$ cogserver
Using default config at /usr/local/etc/cogserver.conf
Listening on port 17001
```
and
```
$ telnet localhost 17001
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
opencog> py
Entering python shell; use ^D or a single . on a line by itself to exit.
py> print atomspace
```
The above has to do a full initialization of python, since no one has done it before.  But also, because it has done the full initialization, it is also responsible for finalization.  

In the past, I think we finalized python when the shell was closed, but clearly that won't work any more.  Also, not that the cogserver allows multiple open python shells, all using the same atomspace!  So everything has to be multi-threaded.  Note also that this atomspace has to be the same one as scheme, so if you `telnet localhost 17001` and start both a scheme shell and a python shell, both work on the same atomspace (no matter which shell you opened first).

Over the years ... something like 6 or 8 or 10 people touched this code, and every time they fixed one bug, they broke something else, so the fixes kept coming in until it turned into spaghetti code.   In particular, I think that forcing python users to explicitly call intialize and finalize is insane:
```
from opencog.utilities import initialize_opencog, finalize_opencog

class AtomTest(TestCase):

    def setUp(self):
        self.space = AtomSpace()
        initialize_opencog(self.space)

    def tearDown(self):
        finalize_opencog()
        del self.space
``` 

That's just plain-old insane; everything should be fully and completely initialized by the time you get done saying `import opencog` ... but the insanity is due to the spaghetti code, no one could figure out how to make everything work automatically, no matter which of the above scenarios were used.

I guess there is even a fourth case:
```
$ python3
Type "help", "copyright", "credits" or "license" for more information.
>>> from opencog import cogserver
>>> start_cogserver()
```
and then `telnet 17001` and open a python shell: this should use the same atomspace (and also not crash, hang, etc.)

Ideally, all of these cases work, and no user is ever forced to call `from opencog.utilities import initialize_opencog, finalize_opencog` (these should not even exist) and no user should ever be forced to call `from opencog.type_constructors import *` and other such nonsense...

Reply:
well, and two more use-cases: from the bash prompt:
```
$ guile
GNU Guile 2.9.2
Enter `,help' for help.
scheme@(guile-user)> (use-modules (opencog) (opencog python))
scheme@(guile-user)> (python-eval "print ('hello! ' + str(2+2))")
```
There's also
```
scheme@(guile-user)> ,describe python-call-with-as
``` 
which reminds me that `set_type_ctor_atomspace` is yet anther bit of python-to-atomspace crazy sphaghetti code that needs to be fixed. You don't have to fix all these issues now, its just that ... they are the ones that keep coming back and biting.

Reply:
And recursively:
```
$ guile
GNU Guile 2.9.2
Enter `,help' for help.
scheme@(guile-user)> (use-modules (opencog) (opencog python))
(python-eval "
      from opencog.atomspace import AtomSpace, TruthValue
      from opencog.atomspace import types

      def foo(asp):
          TV = TruthValue(0.42, 0.69)
          asp.add_node(types.ConceptNode, 'Apple', TV)
          return TV
      ")

(cog-evaluate!
        (Evaluation
                (GroundedPredicate "py:foo")
                (List (Concept "fruit") (Concept "banana"))))
```
and other mixtures, e.g. recursive ones.  In the above example, python is not initialized until the `(use-modules (opencog python))` (which does not have to be at the beginning, but can come much later) and after that, python has to be fully initialized, and ready-to-use, after that guile module is loaded.

Reply:
external code may call global_python_finalize and not to call Py_Finalize, then it would be possible to call intialize_opencog again, harmlessly

Reply:
OK. but that's insane. The only "external code" that we have is the cogserver. I did not look to see if the cogserver calls Py_Finalize or not. Sure, some day in the future, there might be some other "external code" besides the cogserver ... but at the current rate of progress, that future is very far away.  Right now, the best thing to do si to make sure that the cogserver is not insane, and that all of these variations work correctly.

(Last week, I did a fairly deep-cleaning of the cogserver. It is now simpler and more modular than before. I did not touch the python inside of it.)

Reply:
How is that? We have a lot of code in python

Reply:
? Why is the tone so argumentative?  The python code was abandon-ware, until you guys showed up, and fixed a lot of bugs .. for which I am grateful.  Overall, it's still buggy and unstable as hell, failing half the time ... the python bindings are orphans, no one wants to take care of them.   I'm just trying to encourage you to be professional about this, and fix all the remaining bugs and crashes and misbehaviors ... just slapping on more ugly band-aids to something rickety and mis-designed .. just extends the ongoing issues with this code.

Reply:
> Overall, it's still buggy and unstable as hell, failing half the time ...

@linas , I see only three bugs on Python API in list of issues: one of them seems already fixed, other two are fixed by this PR. If you know some critical problems for Python API, please raise issues for them.

On this PR: I agree to make behavior for external and internal initialization consistent, i.e. forbid re-initialization for both internal and external cases. I think it will simplifies code.

@noskill, have you tried it? Does any unit-test fail after this change? Do we have major cases which don't work without re-initialization?

Reply:
@vsbogd cognets use it, i think old vqa project too. I use this initialize_opencog, finalize_opencog to set temporary atomspaces as the current default. Perhaps initialize_opencog does too much, we can try to remove initialize_opencog from python api and give a user set_current_atomspace instead.

Reply:
@linas I made separate issue to discuss python refactoring https://github.com/opencog/atomspace/issues/2391